src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -467,63 +479,68 @@ private void ReceiveBlob(byte[] buffer)
 
             if (_framing == Framing.Unknown)
             {
-                _framing = DetectFraming(buffer, readBytes);
+                _framing = DetectFraming(buffer);
             }
 
-            int restBytes = GetRemainingFrameSize(buffer, 0, readBytes);
+            int frameSize = GetFrameSize(new Span<byte>(buffer, 0, readBytes));
+            int restBytes = frameSize - readBytes;
 
-            if (restBytes < 0)
+            if (frameSize < 0)
             {
                 throw new IOException(SR.net_ssl_io_frame);
             }
 
-            if (restBytes == 0)
-            {
-                // EOF received
-                throw new AuthenticationException(SR.net_auth_eof, null);
-            }
+            buffer = EnsureBufferSize(buffer, readBytes, frameSize);
+
+            restBytes = FixedSizeReader.ReadPacket(_innerStream, buffer, readBytes, frameSize - readBytes);
+
+            SendBlob(new ReadOnlySpan<byte>(buffer, 0, readBytes + restBytes));
+        }
+
+        private static async ValueTask<int> ReadAtLeastAsync(Stream stream, Memory<byte> buffer, int minReadBytes, CancellationToken cancellationToken = default)
```

Comment:
This appears to be used in only one place?  It'd be better to manually inline it into that caller.  Otherwise, if this yields, we're going to pay for the async state machine for both it and its caller rather than just for its caller.

Reply:
It turned out I failed to push one more update to use the `TReadAdapter` as we discussed offline while back @stephentoub. With that, observed Assets with sync/async are gone so as `ReadAtLeastAsync()`. Sorry for the turn.